source/ualloc.c

```diff
@@ -27,12 +27,44 @@ extern int printf(const char *, ...);
 
 extern void * volatile mbed_sbrk_ptr;
 
-// Set debug level to 0 until non-allocating printf is available
-const UAllocDebug_t ualloc_debug_level = UALLOC_DEBUG_NONE;//(DEBUG?UALLOC_DEBUG_MAX:UALLOC_DEBUG_NONE);
+/*****************************************************************************************
+UALLOC_DEBUG_LOG is effectively reserved for tracing memory allocations
+Tracing must be enabled via "yotta config":
+{
+    "debug": {
+        "options": {
+            "memory-trace": true
+        }
+    }
+}
+[TODO]: ualloc_debug_level probably needs better control with "yotta config"
+*****************************************************************************************/
+#ifdef YOTTA_CFG_DEBUG_OPTIONS_MEMORY_TRACE
+const UAllocDebug_t ualloc_debug_level = UALLOC_DEBUG_LOG;
+#else
+const UAllocDebug_t ualloc_debug_level = UALLOC_DEBUG_NONE;
+#endif
 
 // Debug characters
 const char ua_chars[] = "NFEWIL";
 
+/*****************************************************************************************
+The purpose of "prevent_tracing" below is twofold:
+
+1. prevent infinite loops (mbed_ualloc_internal -> ualloc_debug -> printf ->
+   mbed_ualloc_internal -> ualloc_debug -> printf...)
+2. prevent a possible scenario when an interrupt occurs while ualloc_debug
+   is printing debug information and the interrupt also calls 'mbed_ualloc' or
+   another memory operation that would result in ualloc_debug being called,
+   which would in turn result in garbled output. By using "prevent_tracing",
```

Comment:
Actually, I think this is already re-entrant, since it always starts with a known prefix that isn't repeated im the rest of the message. I can't remove `prevent_tracing` yet tohugh, because of printf.